### ISSUE TITLE
Implement vgc::ui module containing QtWidgets-independent UI code

### ISSUE THREAD

**Issue by dalboris** [vgc/vgc#252]
Instead of our current vgc::widgets module based on QtWidgets, we will slowly transition to a vgc::ui module which does not depend on QtWidgets. See #200.

**Comment by dalboris** [comment]
Related PRs which have been assigned the wrong GitHub issue: #273, #274, #275

**Comment by dalboris** [comment]
Note that in `vgc::ui::Widget`, we are planning to be a little more low-level than Qt. For example, instead of having the events automatically propagate to children, I am planning to require the Widget subclasses themselves to call, for example, the `onMouseDown()` callback of their children to perform the propagation (potentially such call to `onMouseDown()` may be indirect, e.g., by calling the base implementation `Widget::onMouseDown()`).

This means slightly more implementation work by subclasses, and more "risks to get it wrong", but the benefits is that it provides much greater customizability (subclasses can even customize how event capturing/bubbling happen), and no hidden/confusing "magic" happening under the hood. For example, things that annoyed in Qt are:

1. Children widgets where always painted after parent widgets. This means that it is not possible for the parent widget to draw thing "on top" of its children (e.g., highlighting, help bubbles, etc.). Such drawovers had to be another sibling widget, covering the other widgets, but then such widget would prevent sibling widgets below it to capture mouse events.

2. In Qt, it is impossible (as far as I know) to capture mouse events before it is propagated to its children. I would like the flexibility for the parent widget to stop the propagation, either up or down. In other words, in Qt, the "bubble" phase of event propagation is somewhat customizable, but not the "capture" phase (borrowing from [Javascript terminology](https://javascript.info/bubbling-and-capturing)). For example, when implementing a resize handle between two widgets, if the mouse is *within some threshold* of the handle, you want to pass the mouse event to the handle (not the the widget under the mouse), even though the mouse is outside the geometry() of the handle. Artificially increasing the handle's "geometry" to overlap with widget on each side causes other problems.

3. More generally, decorellating drawn geometry with mouse event propagation makes it easier to implement shadows, highlighting effects, and overall more flexible/animatable UI.

Finally, note that giving more responsibilities to subclasses means a simpler API design overall, with less obscure "manager" classes doing the work for you. It's a bit more akin to Python philosophy: "explicit is better than implicit".

**Comment by dalboris** [comment]
Note regarding using floats vs. doubles in `vgc::ui`. I decided to use floats for the following reasons:

1. The values that we use in `vgc::ui` are well within the range of floats. Indeed, the sizes/positions of widgets are below < 10k, and even if a widget has a size of 10,000 pixels, such size still has around 10^-3 resolution power (i.e, a float can differentiate between 10000.001 and 10000.002, see below), which is more than enough for UI purposes.

2. There is normally no need for any complex computations that would benefit from higher precision during temporary steps in the computation (like we may need in other modules, e.g., for segment/segment intersection tests when computing a planar graph). In the UI, most computation is just very stable and well-conditioned additions, substractions, comparisons, or multiplication with rotation/scale/translation matrices, not very prone to rounding errors.

3. Rendering the UI eventually requires to convert the values to floats, since this is what most graphics API require, and GPU are optimized for. So using floats in the `vgc::ui` module avoids the performance costs of doing double->float conversions, the memory/performance cost of storing doubles rather than floats (cache miss), but more importantly makes implementing the `onPaint()` functions more concise and readable, since there is no need to do any casting.

Testing float resolution around 10,000:

```
#include <iostream>
#include <iomanip>
int main()
{
    std::cout << std::fixed << std::setprecision(10) << 10000.00000f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00001f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00010f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00020f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00030f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00040f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00050f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00060f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00070f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00080f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00090f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00100f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00200f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00300f << std::endl;
    std::cout << std::fixed << std::setprecision(10) << 10000.00400f << std::endl;
    return 0;
}
```

Output:

```
10000.0000000000
10000.0000000000
10000.0000000000
10000.0000000000
10000.0000000000
10000.0000000000
10000.0009765625
10000.0009765625
10000.0009765625
10000.0009765625
10000.0009765625
10000.0009765625
10000.0019531250
10000.0029296875
10000.0039062500
```

**Comment by dalboris** [comment]
The `vgc::ui` module is now in a reasonable state already, so closing this "create vgc::ui module" issue. Future improvements to `vgc::ui` can reference the following new issue, or other more specific issues:

- https://github.com/vgc/vgc/issues/994
